### PR TITLE
[WIP][help needed] 3D Depthwise Conv

### PR DESCRIPTION
**Pull request by bjuncek**
Refactoring and cleaning up of the #27544

Now builds, but there is a dimension mismatch somewhere. Whilst I slowly try to debug it, any help is very much appreciated,

```
In [3]: conv = nn.Conv3d(2, 2, kernel_size=3, groups=2).to("cuda", torch.float)
   ...: # Run example (this throws errors)
   ...: conv(torch.rand(1, 2, 16, 112, 112).cuda())
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
      1 conv = nn.Conv3d(2, 2, kernel_size=3, groups=2).to("cuda", torch.float)
      2 # Run example (this throws errors)
----> 3 conv(torch.rand(1, 2, 16, 112, 112).cuda())

~/torch_projects/pytorch_bjuncek/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    530             result = self._slow_forward(*input, **kwargs)
    531         else:
--> 532             result = self.forward(*input, **kwargs)
    533         for hook in self._forward_hooks.values():
    534             hook_result = hook(self, input, result)

~/torch_projects/pytorch_bjuncek/torch/nn/modules/conv.py in forward(self, input)
    478                             self.dilation, self.groups)
    479         return F.conv3d(input, self.weight, self.bias, self.stride,
--> 480                         self.padding, self.dilation, self.groups)
    481
    482

IndexError: Dimension out of range (expected to be in range of [-3, 2], but got 3)
```

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 9fcd1db7:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9fcd1db781556202b244cb4331b143b5750eca99).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4140656) pytorch_linux_xenial_py3_6_gcc5_4_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80757311) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80757311))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/scripts/binary_checkout.sh 
Auto-merging .circleci/scripts/binary_checkout.sh 
CONFLICT (add/add): Merge conflict in .circleci/generate_config_yml.py 
Auto-merging .circleci/generate_config_yml.py 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/caffe2_build_definitions.py 
Auto-merging .circleci/cimodel/data/caffe2_build_definitions.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/binary_build_definitions.py 
Auto-merging .circleci/cimodel/data/binary_build_definitions.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4140655) pytorch_xla_linux_xenial_py3_6_clang7_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80757343) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80757343))
<details>
<summary>
<code>Jan 06 17:33:13 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff</code>
</summary>

```
Jan 06 17:33:13 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 06 17:33:13 1 out of 1 hunk FAILED 
Jan 06 17:33:13 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 06 17:33:13 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 06 17:33:13 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Jan 06 17:33:13 Traceback (most recent call last): 
Jan 06 17:33:13   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 06 17:33:13     patch_repo(args) 
Jan 06 17:33:13   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 06 17:33:13     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 06 17:33:13 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Jan 06 17:33:13 + cleanup 
Jan 06 17:33:13 + retcode=1 
Jan 06 17:33:13 + set +x 
Jan 06 17:33:13 =================== sccache compilation log =================== 
Jan 06 17:33:13 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 06 17:33:13 Compile requests                 1 
Jan 06 17:33:13 Compile requests executed        0 
Jan 06 17:33:13 Cache hits                       0 
Jan 06 17:33:13 Cache misses                     0 
Jan 06 17:33:13 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/cuda/VolumetricChannelwiseConvolution.cu [hunk 1]
```diff
@@ -0,0 +1,884 @@
...
+    
+    TORCH_CHECK(
+        (input_.ndimension() == 4),
+        "non-empty 5D tensor expected for input");
+
+    // We should allocate the tensor somewhere here
+    auto options = input_.options();
+    auto output_size = internal::get_output_size<4>(
+      input_, kernel_size, stride_size, pad_size, dilation_size);
+    output = at::empty(output_size, options);
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/VolumetricChannelwiseConvolution.cu]: This looks wrong

File: aten/src/ATen/native/cuda/VolumetricChannelwiseConvolution.cu [hunk 2]
```diff
@@ -0,0 +1,884 @@
...
+    //int64_t istrideD, istrideT, istrideH, istrideW;
+    int64_t totalZ;
+    
+    sizeB  = input_.size(0);
+    sizeD = input_.size(1);
+    isizeT = input_.size(2);
+    isizeH = input_.size(3);
+    isizeW = input_.size(4);
+
+    const Tensor& input = input_.contiguous();
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/VolumetricChannelwiseConvolution.cu]: This also looks wrong

Comment (ezyang) [reply aten/src/ATen/native/cuda/VolumetricChannelwiseConvolution.cu]: Actually const-reference lifetime extension will save you here, I guess

**Comment by bjuncek** [comment]
I've pushed the latest version. 
The code now runs (so the size issue is fixed (in an ugly manner)) but the results are wrong.

My hypothesis is that the indexes I'm passing into the actual kernel are wrong, specifically that one of the 
```
bias_data,
bias_flag, // if true, bias is not null
totalZ, // totalElements
```
is calculated incorrectly w.r.t. how it's expected to be computed. Will continue examination tomorrow.

File: aten/src/ATen/native/cuda/VolumetricChannelwiseConvolution.cu [hunk 3]
```diff
@@ -0,0 +1,859 @@
...
+    int outputTime = (length + 2 * padT - (dilationT * (kT - 1) + 1)) / dT + 1;
+    int outputHeight = (height + 2 * padH - (dilationH * (kH - 1) + 1)) / dH + 1;
+    int outputWidth = (width + 2 * padW - (dilationW * (kW - 1) + 1)) / dW + 1;
+    int outputChannels = weight.size(0);
+  
+    // Pass empty contiguous tensor and resize it here
+    output.resize_({batchSize, outputChannels, outputTime, outputHeight, outputWidth});
+
+    // make sure input is contiguous
+    const Tensor& input = input_.contiguous();
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/VolumetricChannelwiseConvolution.cu]: The use of const reference lifetime extension here is gratuitous, prefer not to do it

**Comment by bjuncek** [comment]
Latest update:

when built with DEBUG=1 tests failing with:
```
ile "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 153, in call_helper
    f(self, *args, dtype=dtype)
  File "test/test_nn.py", line 4217, in test_Conv3d_channelwise_naive_groups_cuda
    output = m(i)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/torch/nn/modules/module.py", line 540, in __call__
    result = self.forward(*input, **kwargs)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/torch/nn/modules/conv.py", line 480, in forward
    self.padding, self.dilation, self.groups)
RuntimeError: CUDA error: too many resources requested for launch

======================================================================
ERROR: test_Conv3d_channelwise_naive_groups_cuda (__main__.TestNN) (dtype=torch.float64)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 153, in call_helper
    f(self, *args, dtype=dtype)
  File "test/test_nn.py", line 4217, in test_Conv3d_channelwise_naive_groups_cuda
    output = m(i)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/torch/nn/modules/module.py", line 540, in __call__
    result = self.forward(*input, **kwargs)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/torch/nn/modules/conv.py", line 480, in forward
    self.padding, self.dilation, self.groups)
RuntimeError: CUDA error: too many resources requested for launch

======================================================================
ERROR: test_Conv3d_channelwise_naive_groups_cuda (__main__.TestNN) (dtype=torch.float16)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 153, in call_helper
    f(self, *args, dtype=dtype)
  File "test/test_nn.py", line 4217, in test_Conv3d_channelwise_naive_groups_cuda
    output = m(i)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/torch/nn/modules/module.py", line 540, in __call__
    result = self.forward(*input, **kwargs)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/torch/nn/modules/conv.py", line 480, in forward
    self.padding, self.dilation, self.groups)
RuntimeError: CUDA error: too many resources requested for launch

======================================================================
FAIL: test_Conv3d_channelwise_naive_groups_cuda (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 678, in wrapper
    method(*args, **kwargs)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 678, in wrapper
    method(*args, **kwargs)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 534, in __exit__
    self.name, after - before, i))
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 890, in assertEqual
    super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
AssertionError: 15872 not less than or equal to 1e-05 : __main__.TestNN.test_Conv3d_channelwise_naive_groups_cuda leaked 15872 bytes CUDA memory on device 0

----------------------------------------------------------------------
Ran 1 test in 10.250s
```


While on the "non debug" build, I get the following error:

```

======================================================================
FAIL: test_Conv3d_channelwise_naive_groups_cuda (__main__.TestNN) (dtype=torch.float16)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 153, in call_helper
    f(self, *args, dtype=dtype)
  File "test/test_nn.py", line 4243, in test_Conv3d_channelwise_naive_groups_cuda
    prec=dtype2prec[dtype])
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 859, in assertEqual
    assertTensorsEqual(x, y)
  File "/private/home/bkorbar/torch_projects/pytorch_bjuncek/test/common_utils.py", line 829, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(0.1853, device='cuda:0', dtype=torch.float16, grad_fn=<MaxBackward1>) not less than or equal to 0.01 :
```

which points towards the fact that I messed something up with the kernel port. Will work on it more today/tomorrow.

**Pull request closed without merging**